core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -428,20 +465,84 @@ public HashDistributionTaskSource(
+                ImmutableListMultimap.Builder<PlanNodeId, ExchangeSourceHandle> exchangeSources = ImmutableListMultimap.builder();
+                long splitsWeight = 0;
+                long exchangeSourcesSize = 0;
+
+                for (TaskDescriptor task : groupTasks) {
+                    ListMultimap<PlanNodeId, Split> taskSplits = task.getSplits();
+                    ListMultimap<PlanNodeId, ExchangeSourceHandle> taskExchangeSources = task.getExchangeSourceHandles();
+                    long taskSplitWeight = taskSplits.values().stream().mapToLong(split -> split.getSplitWeight().getRawValue()).sum();
+                    long taskExchangeSourcesSize = taskExchangeSources.values().stream().mapToLong(this::sourceHandleSize).sum();
+
+                    if ((splitsWeight > 0 || exchangeSourcesSize > 0)
+                            && ((splitsWeight + taskSplitWeight) > targetPartitionSplitWeight || (exchangeSourcesSize + taskExchangeSourcesSize + replicatedExchangeSourcesSize) > targetPartitionSourceSizeInBytes)) {
+                        exchangeSources.putAll(replicatedExchangeSourceHandles); // add replicated exchanges
+                        joinedTasks.add(new TaskDescriptor(taskPartitionId++, splits.build(), exchangeSources.build(), groupNodeRequirements));
+                        splits = ImmutableListMultimap.builder();
+                        exchangeSources = ImmutableListMultimap.builder();
+                        splitsWeight = 0;
+                        exchangeSourcesSize = 0;
+                    }
+
+                    splits.putAll(taskSplits);
+                    exchangeSources.putAll(taskExchangeSources);
+                    splitsWeight += taskSplitWeight;
+                    exchangeSourcesSize += taskExchangeSourcesSize;
+                }
+
+                ImmutableListMultimap<PlanNodeId, Split> remainderSplits = splits.build();
+                ImmutableListMultimap<PlanNodeId, ExchangeSourceHandle> remainderExchangeSources = exchangeSources.build();
+                if (!remainderSplits.isEmpty() || !remainderExchangeSources.isEmpty()) {
+                    remainderExchangeSources = ImmutableListMultimap.<PlanNodeId, ExchangeSourceHandle>builder()
+                            .putAll(remainderExchangeSources)
+                            .putAll(replicatedExchangeSourceHandles) // add replicated exchanges
+                            .build();
+                    joinedTasks.add(new TaskDescriptor(taskPartitionId++, remainderSplits, remainderExchangeSources, groupNodeRequirements));
+                }
+            }
+            return joinedTasks.build();
+        }
+
+        private long sourceHandleSize(ExchangeSourceHandle handle)
+        {
+            Exchange exchange = exchangeForHandle.get(handle);
+            ExchangeSourceStatistics exchangeSourceStatistics = exchange.getExchangeSourceStatistics(handle);
```

Comment:
I'm thinking if it isn't a design mistake to have the `getExchangeSourceStatistics` method under the `Exchange` interface instead of the `ExchangeManager`. Maintaining these mappings could be costly if the number of partitions is high and it seems to be completely unnecessary. Do you think it is worth fixing it? (could be done as a follow up PR)

Reply:
It could be move there. 
But then so should `ExchangeSourceSplitter split(ExchangeSourceHandle handle, long targetSizeInBytes);`, right?

Reply:
leaving as a followup